tests/test_lfs.py

```diff
@@ -74,3 +85,197 @@ def test_lfs(tmp_dir: TmpDir, scm: Git, lfs_objects: TmpDir):
+                                        "href": lfs_object_url,
+                                    }
+                                },
+                            }
+                        ],
+                        "hash_algo": "sha256",
+                    },
+                )
+            ),
+        )
+        m.get(
+            lfs_object_url,
+            callback=recorder.record(
+                CallbackResult(
+                    status=HTTPStatus.TOO_MANY_REQUESTS,
+                    headers=rate_limit_header(),
+                    reason="Too many requests",
+                )
+            ),
+        )
+        m.get(
+            lfs_object_url,
+            callback=recorder.record(
+                CallbackResult(
+                    status=HTTPStatus.OK,
+                    body="lfs data",
+                )
+            ),
+        )
+
+        client.download(storage, [Pointer(oid=FOO_OID, size=FOO_SIZE)])
+
+        results = recorder[lfs_batch_url]
+        assert [r.status for r in results] == [429, 200]
+
+        results = recorder[lfs_object_url]
+        assert [r.status for r in results] == [429, 200]
+
+
+@pytest.mark.parametrize(
+    "rate_limit_header",
+    [
+        lambda: {"Retry-After": "1"},
+        lambda: {"RateLimit-Reset": f"{int(time()) + 1}"},
+        lambda: {"X-RateLimit-Reset": f"{int(time()) + 1}"},
+    ],
+)
+def test_rate_limit_max_retries_batch(
+    storage: LFSStorage, rate_limit_header: Callable[[], dict[str, str]]
+):
+    client = LFSClient.from_git_url("http://git.example.com/namespace/project.git")
+    recorder = CallbackResultRecorder()
+
+    with aioresponses() as m:
+        m.post(
+            f"{client.url}/objects/batch",
+            callback=recorder.record(
+                CallbackResult(
+                    status=HTTPStatus.TOO_MANY_REQUESTS,
+                    headers=rate_limit_header(),
+                    reason="Too many requests",
+                )
+            ),
+            repeat=True,
+        )
+
+        with pytest.raises(ClientResponseError, match="Too many requests"):
+            client.download(storage, [Pointer(oid=FOO_OID, size=FOO_SIZE)])
+
+        results = recorder[f"{client.url}/objects/batch"]
+        assert [r.status for r in results] == [429] * 5
+
+
+@pytest.mark.parametrize(
+    "rate_limit_header",
+    [
+        lambda: {"Retry-After": "1"},
+        lambda: {"RateLimit-Reset": f"{int(time()) + 1}"},
+        lambda: {"X-RateLimit-Reset": f"{int(time()) + 1}"},
+    ],
+)
+def test_rate_limit_max_retries_objects(
+    storage: LFSStorage, rate_limit_header: Callable[[], dict[str, str]]
+):
+    client = LFSClient.from_git_url("http://git.example.com/namespace/project.git")
+    lfs_batch_url = f"{client.url}/objects/batch"
+    lfs_object_url = f"http://git-lfs.example.com/{FOO_OID}"
+    recorder = CallbackResultRecorder()
+
+    with aioresponses() as m:
+        m.post(
```

Comment:
ok, now that I see the difference between tests, can we move each of these `m.post`, `m.get` blocks into a separate function and reuse them across tests? so that each test is small in size (and probably will be move descriptive)? wdyt?

Reply:
Good idea! :+1: I've pushed a commit with refactored tests. WDYT?